### PR TITLE
in-place threshold, sqrt stability fixes, faster logsoftmax

### PR DESCRIPTION
**Pull request by soumith**
This PR has to wait for https://github.com/torch/cutorch/pull/120 to merge, as the LogSoftMax depends on some code from that.

### CODE REVIEW

File: LogSoftMax.cu [hunk 1]
```diff
@@ -1,173 +1,297 @@
...
-    output_k[i] = input_k[i] - logsum_k;
+template <typename Reduction>
+__device__ __forceinline__ float
+blockReduce(float* smem, float val,
+            const Reduction& r,
+            float defaultVal) {
+  return blockReduce<Reduction, NoFinal>(smem, val, r, defaultVal, NoFinal());
 }
 
+template <typename Reduction, int ILP>
```
Comment (nicholas-leonard) [inline LogSoftMax.cu]: ILP?

Comment (nicholas-leonard) [inline LogSoftMax.cu]: http://en.wikipedia.org/wiki/Instruction-level_parallelism

File: LogSoftMax.cu [hunk 2]
```diff
@@ -1,173 +1,297 @@
...
+                                       float *output,
+                                       float *gradOutput,
+                                       int classes) {
+  extern __shared__ float buffer[];
+  gradInput += blockIdx.x * classes;
+  output += blockIdx.x * classes;
+  gradOutput += blockIdx.x * classes;
+
+  float threadSum =
+    ilpReduce<SumFloat, 4>(gradOutput, classes, SumFloat(), 0.0f);
```
Comment (wickedfoo) [inline LogSoftMax.cu]: <SumFloat, 4> should be <SumFloat, ILP> I think. Harmless (not a correctness bug) but I think an ILP level of 2 was optimal when I tested it, which is what this is instantiated at. This was probably left over from a previous version.

File: LogSoftMax.cu (hunk 1 above)
Comment (wickedfoo) [inline LogSoftMax.cu]: Look up Vasily Volkov, "Better Performance at Lower Occupancy" for a general overview of the strategy.

File: LogSoftMax.cu [hunk 3]
```diff
@@ -1,173 +1,297 @@
...
+__device__ __forceinline__ float
+blockReduce(float* smem, float val,
+            const Reduction& r,
+            float defaultVal) {
+  return blockReduce<Reduction, NoFinal>(smem, val, r, defaultVal, NoFinal());
 }
 
+template <typename Reduction, int ILP>
+__device__ __forceinline__ float
+ilpReduce(float* data,
```
Comment (dominikgrewe) [inline LogSoftMax.cu]: Would `threadReduce` be a more descriptive name, considering that you also have `blockReduce`?

Comment (wickedfoo) [inline LogSoftMax.cu]: yes, sounds good

**Comment by wickedfoo** [comment]
this was the LSM speedup:
(minibatch size, # classes): old time (ms), new time (ms), speedup

forward 1 10000: 0.15497207641602 0.082015991210938 1.88953x
backward 1 10000: 0.1368522644043 0.076055526733398 1.79937x
forward 1 100000: 0.90694427490234 0.16188621520996 5.60236x
backward 1 100000: 1.1420249938965 0.52905082702637 2.15863x
forward 1 1000000: 12.041807174683 1.5771389007568 7.63522x
backward 1 1000000: 10.465145111084 1.1870861053467 8.81583x
forward 16 10000: 0.54001808166504 0.46110153198242 1.17115x
backward 16 10000: 0.55694580078125 0.45490264892578 1.22432x
forward 16 100000: 1.6398429870605 0.6411075592041 2.55783x
backward 16 100000: 1.4832019805908 0.58698654174805 2.52681x
forward 16 1000000: 12.393951416016 2.1309852600098 5.81607x
backward 16 1000000: 10.847091674805 1.9412040710449 5.58782x
forward 32 10000: 0.56195259094238 0.48708915710449 1.1537x
backward 32 10000: 0.55885314941406 0.47206878662109 1.18384x
forward 32 100000: 1.6579627990723 0.78797340393066 2.10408x
backward 32 100000: 1.514196395874 0.75078010559082 2.01683x
forward 32 1000000: 12.70604133606 4.0738582611084 3.11892x
backward 32 1000000: 11.248111724854 3.6840438842773 3.0532x
forward 64 10000: 0.61988830566406 0.50997734069824 1.21552x
backward 64 10000: 0.57506561279297 0.51021575927734 1.1271x
forward 64 100000: 1.7609596252441 1.0561943054199 1.66727x
backward 64 100000: 1.5981197357178 0.99802017211914 1.60129x
forward 64 1000000: 13.644933700562 6.7241191864014 2.02925x
backward 64 1000000: 12.102842330933 6.3221454620361 1.91436x
forward 128 10000: 0.60892105102539 0.56600570678711 1.07582x
backward 128 10000: 0.59914588928223 0.55813789367676 1.07347x
forward 128 100000: 1.9750595092773 1.5769004821777 1.25249x
backward 128 100000: 1.8389225006104 1.5139579772949 1.21465x
forward 128 1000000: 15.992879867554 11.859178543091 1.34857x
backward 128 1000000: 14.727830886841 11.378049850464 1.29441x

**Comment by soumith** [comment]
coming in another PR

**Pull request closed without merging**